## <div style="text-align: left"> Assigment 2 CT3113</div>
## <div style="text-align: left"> Autor: Leonardo Figuera</div>
<img src="CT3331_tarea_2_para_entregar_el_31_mayo.JPG" alt="Enunciado Tarea 1" style="width: 300px;" align="center"/>

Performing an analysis by inspection of meshes we obtain:

$$\begin{equation}
\begin{pmatrix}
-E\\
E\\
-E3
\end{pmatrix}
=
\begin{pmatrix}
R3+R5 & 0 & R3\\
0 & R4+R6 & R4\\
R3 & R4 & R1+R2+R3+R4
\end{pmatrix}
*
\begin{pmatrix}
I1\\
I2\\
I3
\end{pmatrix}
\end{equation}
$$

Then we proceed to solve with Python:

* First of all, the Numpy library files must be brought into the work environment to operate on vectors and matrices.
* Keep in mind that indexing in Python begins with ***0***, so when calling ***I[0]*** you are calling the first element of the array ***I***

In [1]:
# In this line the numpy library files are brought
import numpy as np

In [2]:
# Definition of values
E = 1.05
E3 = 0.98
R1 = 0.1
R2 = 0.2
R3 = 0.3
R4 = 0.4
R5 = 0.5
R6 = 0.6

# 'np.array' function created arrays to represent matrices and vectors
A = np.array([[R3+R5, 0, R3],
              [0, R4+R6, R4],
              [R3, R4, R1+R2+R3+R4]])

B = np.array([-E,
              E,
              -E3])

# 'np.linalg.solve' function finds solution of the equation of type "Ax = B"
I = np.linalg.solve(A, B)

# Printing of results
print(f"""Calculated mesh currents are:\nI1 = {I[0]:.4f}
I2 = {I[1]:.4f}\nI3 = {I[2]:.4f}\n""")

Calculated mesh currents are:
I1 = -0.7938
I2 = 1.6033
I3 = -1.3832



* Since we obtained negative values for the currents ***I1*** and ***I3***, the assumed direction of the current flow was wrong.

The circuit diagram, with the currents in the correct directions, is:

<img src="DIAGRAMA_T2_BIEN.PNG" alt="Drawing" style="width: 300px;" align="center"/>

Once the mesh currents are obtained, we can calculate the voltage drop in the resistors, knowing the equation:
$$\begin{equation} \Delta V = R * I\end{equation}$$
* If more than one mesh current passes through a resistor, the currents will be summed, taking into cosideration their plus or minus sign, depending on whether when passing through the resistance they have the same or different direction.


In [3]:
## First we will change to positive values the currents obtained with  np.absolute function
I = np.absolute(I)

delta_V1 = R1 * I[2]
delta_V2 = R2 * I[2]
delta_V3 = R3 * (I[2]+I[0])
delta_V4 = R4 * (I[1]-I[2])
delta_V5 = R5 * I[0]
delta_V6 = R6 * I[1]

# Printing of results
print(f"""The voltage drops across the resistors are:
delta_V1 = {delta_V1:.4f} \ndelta_V2 = {delta_V2:.4f} \ndelta_V3 = {delta_V3:.4f}
delta_V4 = {delta_V4:.4f} \ndelta_V5 = {delta_V5:.4f} \ndelta_V6 = {delta_V6:.4f}\n""")

The voltage drops across the resistors are:
delta_V1 = 0.1383 
delta_V2 = 0.2766 
delta_V3 = 0.6531
delta_V4 = 0.0880 
delta_V5 = 0.3969 
delta_V6 = 0.9620



The dissipated power of the resistor can be calculated knowing th equation:
$$\begin{equation} P = \Delta V * I \end{equation}$$
* As in the calculation of voltage drop, the current ***I*** corresponds to the sum, or subtraction, of the currents that pass through the resistor depending on its direction in the mesh analysis.


In [5]:
P1 = delta_V1 * I[2]
P2 = delta_V2 * I[2]
P3 = delta_V3 * (I[2]+I[0])
P4 = delta_V4 * (I[1]-I[2])
P5 = delta_V5 * I[0]
P6 = delta_V6 * I[1]

P_total = P1+P2+P3+P4+P5+P6

# En estas lineas se imprimen los resultados obtenidos
print(f"""The dissipated power values are:
P1 = {delta_V1:.4f} \nP2 = {delta_V2:.4f} \nP3 = {delta_V3:.4f}
P4 = {delta_V4:.4f} \nP5 = {delta_V5:.4f} \nP6 = {delta_V6:.4f}\n""")
print(f"The total dissipated power is:\nPtotal = {P_total:.4f}")

The dissipated power values are:
P1 = 0.1383 
P2 = 0.2766 
P3 = 0.6531
P4 = 0.0880 
P5 = 0.3969 
P6 = 0.9620

The total dissipated power is:
Ptotal = 3.8724


Then, the power generated by the sources is calculated by:
$$\begin{equation} P = E * I \end{equation}$$
* If a source is in contact with more than one mesh current, the current use in this calculaiton will be the sum, taking into cosideration their plus or minus sign, depending on whether when passing through these currents have the same or different direction in the mesh analysis.

In [9]:
P_source_E = E * (I[0]+I[1])
P_source_E3 = E3 * I[2]
P_source = P_source_E + P_source_E3

print(f"The total power generated by the sources is:\nP_fuentes = {P_source:.4f}")

The total power generated by the sources is:
P_fuentes = 3.8724


##### Since the power produced by the sources is equal to the power consumed by the resistors...
$$\begin{equation} P_{resistances} = 3.87 \end{equation}$$ $$\begin{equation} P_{sources} = 3.87 \end{equation}$$
##### ...the resolution of the analysis is assumed to be correct.

##### Finally, Kirchoff's law of voltage in each mesh is verified:

$$E = I_{1}(R3+R5) + I_{3}R3$$
$$E = I_{2}(R4+R6) - I_{3}R4$$
$$E3 = I_{1}R3 - I_{2}R4 + I_{3}(R1+R2+R2+R4)$$

In [10]:
Ee1 = E
Es1 = I[0]*(R3+R5) + I[2]*R3
Ee2 = E
Es2 = I[1]*(R4+R6) - I[2]*R4
Ee3 = E3
Es3 = I[0]*R3 - I[1]*R4 + I[2]*(R1+R2+R3+R4)
print(f"En la mesh 1: total_voltage = {Ee1-Es1:.4f}")
print(f"En la mesh 2: total_voltage = {Ee2-Es2:.4f}")
print(f"En la mesh 3: total_voltage = {Ee3-Es3:.4f}")

En la mesh 1: total_voltage = 0.0000
En la mesh 2: total_voltage = 0.0000
En la mesh 3: total_voltage = 0.0000
